### PR TITLE
created /api/articles/:id/comments endpoint (task 6)

### PR DESCRIPTION
**Pull request by Dave-Turnbull**
also refactored some error handling for bad requests, changed 404 response message when no rows found.

### CODE REVIEW

**Comment by Dave-Turnbull** [comment]
I forgot the endpoint.json data, just added it but I've already merged

**Comment by Dave-Turnbull** [comment]
and also to sort by date, sorry!

File: __tests__/database.test.js [hunk 1]
```diff
@@ -59,13 +58,120 @@ describe("GET endpoints", () => {
...
+            article_id: 1
+          }
+        return request(app)
+        .get('/api/articles/1/comments')
+        .expect(200)
+        .then(({body}) => {
+            expect(body.length).toBe(11)
+            body.forEach(article => {
+                expect(article).toMatchObject(matchCommentObject)
+            });
```
Comment (RichardNeat) [inline __tests__/database.test.js:139]: Great test here and well done for recognising that we can be and therefore should be specific with the article_id but generic with the rest 👏

File: __tests__/database.test.js [hunk 2]
```diff
@@ -59,13 +58,120 @@ describe("GET endpoints", () => {
...
+            expect(body.message).toBe('article not found')
+        })
+    })
+    test("Request for existing article with no comments returns 200 and empty array", () => {
+        return request(app)
+        .get('/api/articles/2/comments')
+        .expect(200)
+        .then(({body}) => {
+            expect(body).toEqual([])
+        })
```
Comment (RichardNeat) [inline __tests__/database.test.js:156]: GREAT to see this test. Well done for differentiating between an empty response as correct and an empty response being an error.

File: controllers/index.js [hunk 3]
```diff
@@ -9,13 +15,39 @@ exports.getEndpoints = (req, res, next) => {
...
+    return retrieveArticles()
+    .then(({rows}) => {
+        res.status(200).send(rows)
+    })
+    .catch(next) 
+}
+
+exports.getCommentsByArticleId = (req, res, next) => {
+    const {params} = req
+    return Promise.all([retrieveComments(params.id), retrieveArticlesById(params.id)])
```
Comment (RichardNeat) [inline controllers/index.js:44]: Excellent choice to handle the potential error this way as the response from retrieving the comments isn't a reliable error check anymore.

File: controllers/models/index.js [hunk 4]
```diff
@@ -11,12 +11,45 @@ exports.retrieveEndpoints = () => {
...
-            return Promise.reject({status: 404, message: 'nothing found'})
+            const dataName = result.fields[0].name.match(/^[a-z]+/i)[0]
+            return Promise.reject({status: 404, message: `${dataName} not found`})
         }
         return result
     })
+}
+
+exports.retrieveTopics = () => {
+    let sqlQuery = `SELECT * FROM topics`
```
Comment (RichardNeat) [inline controllers/models/index.js:26]: const > let!

File: controllers/models/index.js [hunk 5]
```diff
@@ -11,12 +11,45 @@ exports.retrieveEndpoints = () => {
...
+    return retrieveData(sqlQuery)
+}
+
+exports.retrieveArticlesById = (id) => {
+    let sqlQuery = `SELECT * FROM articles WHERE article_id=${id}`
```
Comment (RichardNeat) [inline controllers/models/index.js:31]: const > let!

File: __tests__/database.test.js [hunk 6]
```diff
@@ -59,13 +58,120 @@ describe("GET endpoints", () => {
...
+    })
+    test("Request for invalid article ID returns 400 Bad request", () => {
+        return request(app)
+        .get('/api/articles/invalid/comments')
+        .expect(400)
+        .then(({body}) => {
+            expect(body.message).toBe('Bad request')
+        })
+    })
+})
```
Comment (RichardNeat) [inline __tests__/database.test.js:166]: Would be nice to see a test for checking that the comments are served most recent first.

File: controllers/models/index.js [hunk 7]
```diff
@@ -11,12 +11,45 @@ exports.retrieveEndpoints = () => {
...
+}
+
+exports.retrieveArticles = (id) => {
+    //COALESCE converts the value to 0 if its null
+    //AS comment_count sets the column name
+    //CAST(... AS INT) converts to an interger (apparently GROUP BY returns as a VARCHAR)
+    //COUNT(*) counts all the items in the group
+    //GROUP BY creates groups for all comments seperated by article_id
+    //SQL hurts
+    let sqlQuery = `SELECT author, title, articles.article_id, topic, created_at, votes, article_img_url, COALESCE(comments.comment_count, 0) AS comment_count FROM articles
```
Comment (RichardNeat) [inline controllers/models/index.js:43]: Same as above, const > let. Lets make sure we are sticking to good conventions here. Later we might want to allow this endpoint to take in queries from the user meaning we might need to make this query editable and make it a let, but right now we don't need to do that and don't have a test that has driven us to need to make this a let.

**Review by RichardNeat — COMMENTED** [review]
Really nice! Just a couple of little tweaks in comments below.

**Pull request merged**